### PR TITLE
add monitoring controller for kserve and istio metrics

### PR DESCRIPTION
**Pull request by VedantMahabaleshwarkar**
## Description

- Added Kserve InferenceService Controller
- Modified OpenshiftInferenceService Controller
- [Solution Design](https://lucid.app/lucidspark/b3588482-802a-4592-8320-95472475ae88/edit?viewport_loc=-1358%2C-395%2C5760%2C2769%2C0_0&invitationId=inv_2feedb20-0177-4405-a788-5411a16ede14)  

## How Has This Been Tested?

- This PR should be tested along with corresponding PRs to [caikit-tgis-serving](https://github.com/opendatahub-io/caikit-tgis-serving/pull/46) and [odh-manifests](https://github.com/opendatahub-io/odh-manifests/pull/912)
- To test the new controller : follow the instructions [here](https://github.com/VedantMahabaleshwarkar/caikit-tgis-serving/blob/metrics-controller/demo/kserve/Kserve.md) 

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by vaibhavjainwiz** [comment]
I would suggest to make kserve_inferenceservice_controller as in independent controller which registered itself to operator.

- inferenceservice_controller should only listen for event where Annotations["serving.kserve.io/deploymentMode" = true]
- kserve_inferenceservice_controller should only listen for event where Annotations["serving.kserve.io/deploymentMode"] is missing.

**Comment by vaibhavjainwiz** [comment]
IMHO it would make more sense if could rename `inferenceservice_controller_test.go` to `modelMeshInferenceservice_controller_test.go`

**Comment by VedantMahabaleshwarkar** [comment]
@vaibhavjainwiz 
> I would suggest to make kserve_inferenceservice_controller as in independent controller which registered itself to operator.
inferenceservice_controller should only listen for event where Annotations["serving.kserve.io/deploymentMode" = ModelMesh]
kserve_inferenceservice_controller should only listen for event where Annotations["serving.kserve.io/deploymentMode"] is missing.
through that we could control which controller to register with operator by passing some boolean flag. For example : if enableKServeInferenceService = true and enableModelMeshInferenceService = false then we could just register kserve_inferenceservice_controller. 

I had considered this during my initial design but there are a few reasons this approach seems better to me.
- The Isvc CRD is the same across kserve and modelmesh. In the future modelmesh and kserve will be able to be deployed together so it might be more efficient to just have 1 overall reconcile loop. I feel like having 2 controllers watching the same CRD spec might get unnesessarily complicated.
- Right now the reconciliation logic for both cases is completely different. However in the future we may need to do some common things for model serving regardless of the underlying stack. This architecture leaves room for such a future with relative ease

File: controllers/inferenceservice_controller.go [hunk 1]
```diff
@@ -68,20 +59,35 @@ func (r *OpenshiftInferenceServiceReconciler) Reconcile(ctx context.Context, req
 	err := r.Get(ctx, req.NamespacedName, inferenceservice)
 	if err != nil && apierrs.IsNotFound(err) {
 		log.Info("Stop InferenceService reconciliation")
+		// InferenceService not found, so we check for any other inference services that might be using Kserve
+		// If none are found, we delete the common namespace-scoped resources that were created for Kserve Metrics.
+		error := r.DeleteKserveMetricsResourcesIfNoKserveIsvcExists(ctx, req, req.Namespace)
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/inferenceservice_controller.go:64]: The same thing can be accomplished with a finalizer but the overall logic for deletion in the reconcile loop would still stay the same, right? I didn't get why using `finalizers` is necessary/better here. I've made these changes in a way so as to minimize impact to the modelmesh isvc reconciliation workflow. I think we can definitely use finalizers here, but perhaps we can do that in a follow up PR. WDYT?

[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/inferenceservice_controller.go:64]: ack, will change to err1 (although golang didn't throw any warnings here)

File: controllers/inferenceservice_controller.go [hunk 2]
```diff
@@ -68,20 +59,35 @@ func (r *OpenshiftInferenceServiceReconciler) Reconcile(ctx context.Context, req
...
+		if error != nil {
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/inferenceservice_controller.go:65]: ack, handled

File: controllers/kserve_inferenceservice_controller.go [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+func (r *OpenshiftInferenceServiceReconciler) ReconcileKserveInference(ctx context.Context, req ctrl.Request, inferenceService *kservev1beta1.InferenceService) error {
+
+	// Initialize logger format
+	log := r.Log.WithValues("InferenceService", inferenceService.Name, "namespace", inferenceService.Namespace)
+
+	//At least 1 InferenceService exists in the namespace, now ensure common namespace-scope resources exist
+	//Also create the metrics service and servicemonitor with OwnerReferences, as these are not common namespace-scope resources
+
+	log.Info("Verifying that the rolebinding to enable prometheus access exists")
+	err := r.ensurePrometheusRoleBinding(ctx, req, inferenceService.Namespace)
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/kserve_inferenceservice_controller.go:581]: ack

File: controllers/kserve_inferenceservice_controller.go [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+		if apierrs.IsNotFound(err) {
+			istioServiceMonitorFound = false
+		} else {
+			log.Error(err, "Unable to get Istio ServiceMonitor")
+			return err
+		}
+	}
+	if !istioServiceMonitorFound {
+		desiredServiceMonitor := &monitoringv1.ServiceMonitor{
+			TypeMeta: metav1.TypeMeta{
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/kserve_inferenceservice_controller.go:197]: ack

File: controllers/kserve_inferenceservice_controller.go [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+				8086: {Mode: 2},
+				3000: {Mode: 2},
+			},
+		},
+	}
+
+	if err := r.Client.Create(ctx, peerAuth); err != nil {
+		if errors.IsAlreadyExists(err) {
+			log.Info("PeerAuth already exists for target namespace ")
+			if err := r.Client.Update(ctx, peerAuth); err != nil {
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/kserve_inferenceservice_controller.go:305]: ack

File: controllers/kserve_inferenceservice_controller.go [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+
+	for i := len(inferenceServiceList.Items) - 1; i >= 0; i-- {
+		inferenceService := inferenceServiceList.Items[i]
+		if value, exists := inferenceService.Annotations["serving.kserve.io/deploymentMode"]; exists && value == "ModelMesh" {
+			inferenceServiceList.Items = append(inferenceServiceList.Items[:i], inferenceServiceList.Items[i+1:]...)
+		}
+	}
+
+	// If there are no Kserve InferenceServices in the namespace, delete namespace-scoped resources needed for Kserve Metrics
+	if len(inferenceServiceList.Items) == 0 {
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/kserve_inferenceservice_controller.go:486]: ack

File: controllers/inferenceservice_controller.go [hunk 7]
```diff
@@ -90,14 +96,20 @@ func (r *OpenshiftInferenceServiceReconciler) Reconcile(ctx context.Context, req
...
 		Owns(&predictorv1.ServingRuntime{}).
 		Owns(&corev1.Namespace{}).
 		Owns(&routev1.Route{}).
 		Owns(&corev1.ServiceAccount{}).
 		Owns(&corev1.Service{}).
 		Owns(&corev1.Secret{}).
 		Owns(&authv1.ClusterRoleBinding{}).
+		Owns(&istiosecv1beta1.PeerAuthentication{}).
+		Owns(&networkingv1.NetworkPolicy{}).
+		Owns(&corev1.Service{}).
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/inferenceservice_controller.go:109]: ack

File: controllers/inferenceservice_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vaibhavjainwiz) [reply controllers/inferenceservice_controller.go:64]: Please create a github issue for it and add into backlog

File: controllers/kserve_inferenceservice_controller.go [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+	inferenceServiceList := &kservev1beta1.InferenceServiceList{}
+	err := r.List(ctx, inferenceServiceList, client.InNamespace(req.Namespace))
+	if err != nil {
+		log.Error(err, "Unable to get the list of InferenceServices in the namespace. Cannot delete PeerAuthentication and NetworkPolicy.")
+		return err
+	}
+
+	for i := len(inferenceServiceList.Items) - 1; i >= 0; i-- {
+		inferenceService := inferenceServiceList.Items[i]
+		if value, exists := inferenceService.Annotations["serving.kserve.io/deploymentMode"]; exists && value == "ModelMesh" {
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/kserve_inferenceservice_controller.go:480]: ack

File: controllers/inferenceservice_controller.go [hunk 9]
```diff
@@ -41,21 +47,6 @@ type OpenshiftInferenceServiceReconciler struct {
 	MeshDisabled bool
 }
 
-// ClusterRole permissions
```
[the remark this answers is not in the document]
Comment (vaibhavjainwiz) [reply controllers/inferenceservice_controller.go:44]: no.. these should be on controller level.

**Comment by vaibhavjainwiz** [comment]
> @vaibhavjainwiz
> 
> > I would suggest to make kserve_inferenceservice_controller as in independent controller which registered itself to operator.
> > inferenceservice_controller should only listen for event where Annotations["serving.kserve.io/deploymentMode" = ModelMesh]
> > kserve_inferenceservice_controller should only listen for event where Annotations["serving.kserve.io/deploymentMode"] is missing.
> > through that we could control which controller to register with operator by passing some boolean flag. For example : if enableKServeInferenceService = true and enableModelMeshInferenceService = false then we could just register kserve_inferenceservice_controller.
> 
> I had considered this during my initial design but there are a few reasons this approach seems better to me.
> 
> * The Isvc CRD is the same across kserve and modelmesh. In the future modelmesh and kserve will be able to be deployed together so it might be more efficient to just have 1 overall reconcile loop. I feel like having 2 controllers watching the same CRD spec might get unnesessarily complicated.
> * Right now the reconciliation logic for both cases is completely different. However in the future we may need to do some common things for model serving regardless of the underlying stack. This architecture leaves room for such a future with relative ease

- its exactly opposite. Seperate them will reduce the complexity and provide clear segregation.
- In future if some common thing came then we could create a new service for it and call it from both reconciliation.

File: controllers/inferenceservice_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply controllers/inferenceservice_controller.go]: ack

**Comment by heyselbi** [comment]
Passed testing on an OSD cluster with OpenShift 4.12.23
Screenshot from 2023-08-11 12-16-40
59-4f5d-
Screenshot from 2023-08-11 12-17-00
9a3f-7a06e44233aa)

**Review by heyselbi — APPROVED** [review]

**Comment by vaibhavjainwiz** [comment]
@VedantMahabaleshwarkar 
Couple of more review comment.

**[Issue]** : If we create second inference service in same namespace then creation of `PeerAuthentication` resource is failed with below error. You might not see this error in logs but you could debug this back by checking MetricService & MetricServiceMonitor objects are not created for second InferenceService instance.
```
peerauthentications.security.istio.io "default" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update
```

**[Issue]** : With above issue strange thing is that this err/exception is not logged into log message but program silently skip reconcilation of InferenceService object.

**Comment by vaibhavjainwiz** [comment]
For every InferenceService, we are creating following resources :
ensurePrometheusRoleBinding
ensureServiceMeshMemberRollEntry
createOrUpdateIstioTelemetry	
createOrUpdateIstioServiceMonitor
createOrUpdateIstioPodMonitor
createOrUpdatePeerAuth
createOrUpdateNetworkPolicy
createOrUpdateMetricsService
createOrUpdateMetricsServiceMonitor

Among them all resources(except createOrUpdateMetricsService/createOrUpdateMetricsServiceMonitor) are created once for each namespace. Its not a issue or a problem to put these checks in InferenceService but IMHO it make more sense to have those in ServiceRuntime controller because number of instances of ServiceRuntime would be lesser then the instance of InferenceService per namespace. So that there would be less reconciliation of these resource.

Note : Its just my observation. Please ignore it if you feel defer.

File: config/rbac/role.yaml [hunk 10]
```diff
@@ -2,181 +2,207 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
+  creationTimestamp: null
```
[the remark this answers is not in the document]
Comment (VedantMahabaleshwarkar) [reply config/rbac/role.yaml]: Gets autogenerated on `make manifests`. Removed it

**Comment by heyselbi** [comment]
Tested the new bug fix and works fine with 2 isvcs deployed in one namespace. Checked the route inferencing too, both of them work.

Screenshot from 2023-08-14 12-53-07

**Comment by heyselbi** [comment]
Also getting the expected resources:
```
oc get servicemonitor
NAME                           AGE
caikit-example-isvc-metrics    14m
caikit-example-isvc2-metrics   11m
istiod-monitor                 14m
```
and
```
oc get service
caikit-example-isvc                          
caikit-example-isvc-metrics              
caikit-example-isvc-predictor               
caikit-example-isvc-predictor-00001          
caikit-example-isvc-predictor-00001-private   
caikit-example-isvc2                       
caikit-example-isvc2-metrics           
caikit-example-isvc2-predictor          
caikit-example-isvc2-predictor-00001    
```
cc @vaibhavjainwiz

**Review by Xaenalt — APPROVED** [review]

**Comment by Xaenalt** [comment]
Looks like openshift-ci is having trouble, verified with @VedantMahabaleshwarkar that we're ok to merge

**Pull request merged**